### ISSUE TITLE
FFT on CPU noticeably slower than SciPy's FFT

### ISSUE THREAD

**Issue by Edenhofer** [google/jax#7490]
The FFT implementation in JAX seems to be noticeably slower than the one in SciPy even though both use some flavor of the PocketFFT FFT implementation.

```python
from jax.config import config

config.update("jax_enable_x64", True)

import timeit
import numpy as np
from scipy import fft as sp_fft
from jax import numpy as jnp
from jax import jit, random

key = random.PRNGKey(42)
jr = random.normal(key, (2**26, ))
r = np.array(jr)

N_IT = 7

timing = timeit.timeit(lambda: np.fft.fft(r), number=N_IT) / N_IT
print(f"NumPy: {timing} s")
timing = timeit.timeit(lambda: sp_fft.fft(r), number=N_IT) / N_IT
print(f"SciPy: {timing} s")

jax_fft = jnp.fft.fft
timing = timeit.timeit(
    lambda: jax_fft(jr).block_until_ready(), number=N_IT
) / N_IT
print(f"JAX (unjitted): {timing} s")

jax_fft_jit = jit(jax_fft)
jax_fft_jit(jr)  # Warm-up
timing = timeit.timeit(
    lambda: jax_fft_jit(jr).block_until_ready(), number=N_IT
) / N_IT
print(f"JAX (jitted): {timing} s")
```

On an AMD Ryzen 7 4800H (with Radeon Graphics) with JAX 0.2.18 and jaxlib 0.1.69 installed from PyPI, I get the following timings:

```txt
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
NumPy: 3.5420487681424544 s
SciPy: 1.9076589474290293 s
JAX (unjitted): 4.5806920641433475 s
JAX (jitted): 4.63735637514245 s
```

The timings improve if I compile JAX(lib) myself though it still compares unfavorably to SciPy (compiled JAX (unjitted): 3.506302507857202 s).

My hypothesis is that JAX is distributing binaries that are suboptimal for recent AMD CPUs and much more importantly JAX is probably using some outdated version of PocketFFT.

**Comment by hawkinsp** [comment]
The PocketFFT version used by JAX is pretty close to PocketFFT's C++ branch head. There have been no performance-relevant changes since.

One possibility: is the scipy version multithreaded? I didn't plug PocketFFT's threading support into JAX.
Another possibility is the PocketFFT C and C++ branches have different performance.

**Comment by hawkinsp** [comment]
If all are single-threaded, then I'd be tempted to try writing a small C/C++ benchmark that compares the C and C++ implementations of PocketFFT for your specific input sizes. My best guess is this is a PocketFFT C vs C++ difference. Is that something you might be willing to do?

**Comment by Edenhofer** [comment]
Meaning you believe `python3 build/build.py` with `--target_cpu_features native` (which should in turn set `-march=native`) would make JAX's FFT as fast as SciPy's?

The difference in performance between NumPy and SciPy can be, if I am not mistaken, purely explained by this difference in implementation. Personally, I am much more interested in how to get JAX on par with SciPy (both use the C++ implementation AFAIK) and less interested in benchmarking PocketFFT's C vs. C++ implementation.

**Comment by Edenhofer** [comment]
Can you elaborate a little on how this could explain the timings? Since both SciPy and JAX use the C++ flavor of PocketFFT, I don't understand how the C version could be faster. NumPy, which uses the latter, is slower in this benchmark. If you have the hunch that there is a problem in PocketFFT, I am more than happy to contact its author Martin Reinecke. He works next door and would surely be happy to help.

Note, the above timings are representative for any reasonably large (> 8000) length which can be represented as a power of two. I chose to limit the analysis to powers of two since the FFT algorithm should be most efficient for these shapes.

Last but not least, adding `--target_cpu_features native` does not change anything about the here presented results. To be precise, it has no measurable impact on the timings at all.

**Comment by hawkinsp** [comment]
My apologies: I thought scipy was using the C version of PocketFFT.

Actually, it turns out the explanation is much simpler than I thought: we appear to be casting your (real) input to complex and then doing a complex to complex fft, when we should be doing a real to complex fft. This appears to account for most of the time difference:

```
In [11]: jax.make_jaxpr(jnp.fft.fft)(jr)
Out[11]:
{ lambda  ; a.
  let b = convert_element_type[ new_dtype=complex128
                                weak_type=False ] a
      c = fft[ fft_lengths=(67108864,)
               fft_type=FftType.FFT ] b
  in (c,) }

...

# Use an internal API to actually get an RFFT:
In [17]: %timeit jax.lax.fft(jr, jax.lib.xla_client.FftType.RFFT, (jr.shape[0],)).block_until_ready()
2.33 s ± 24 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [18]: %timeit sp_fft.fft(r)
2.46 s ± 20.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

If you cast the scipy input to a complex value, it will be much closer to the JAX timing. This should be easy to fix...

**Comment by shoyer** [comment]
We can probably speed this up in JAX, but I'm not sure it's absolutely essential. The best optimizations here probably can't match what you can realize using real-valued FFTs in your own code.

**Comment by Edenhofer** [comment]
Ah, that makes perfect sense! Thanks for the explanation!

Personally, I think it would be nice to have some kind of dtype check in JAX's FFT. Isn't it possible to copy the respective lines from SciPy?

Furthermore, while the rfft does indeed bring JAX to the performance of SciPy, the poor performance seems to persist for `np.complex128`:
```python
for dtp in (np.complex128, np.float64):
    for p in range(22, 25):
        print(f"POWER {p}; DATA-TYPE {dtp}")
        jr = random.normal(key, (2**p, ), dtype=dtp)
        r = np.array(jr)

        N_IT = 7

        timing = timeit.timeit(lambda: np.fft.fft(r), number=N_IT) / N_IT
        print(f"NumPy: {timing}s")
        timing = timeit.timeit(lambda: sp_fft.fft(r), number=N_IT) / N_IT
        print(f"SciPy: {timing}s")

        jax_fft = jnp.fft.rfft if dtp == np.float64 else jnp.fft.fft
        timing = timeit.timeit(
            lambda: jax_fft(jr).block_until_ready(), number=N_IT
        ) / N_IT
        print(f"JAX (unjitted): {timing}s")
```

```text
POWER 22; DATA-TYPE <class 'numpy.complex128'>
NumPy: 0.1732045361428131s
SciPy: 0.13564815842853153s
JAX (unjitted): 0.24770665457149985s
POWER 23; DATA-TYPE <class 'numpy.complex128'>
NumPy: 0.34857762385711666s
SciPy: 0.2921157494285061s
JAX (unjitted): 0.625066616285559s
POWER 24; DATA-TYPE <class 'numpy.complex128'>
NumPy: 0.7146986538570802s
SciPy: 0.5851213647144634s
JAX (unjitted): 1.2242363377143712s
POWER 22; DATA-TYPE <class 'numpy.float64'>
NumPy: 0.17276582571425284s
SciPy: 0.10931958271430631s
JAX (unjitted): 0.1242849888570033s
POWER 23; DATA-TYPE <class 'numpy.float64'>
NumPy: 0.34794485971438655s
SciPy: 0.22209468514276004s
JAX (unjitted): 0.2466219107141764s
POWER 24; DATA-TYPE <class 'numpy.float64'>
NumPy: 0.714676908142792s
SciPy: 0.4494652548572568s
JAX (unjitted): 0.5015757604284902s
```

Speaking about real-valued FFTs, I would be incredibly interested in Hartley transformation. Is this something of interest to JAX as well? There is a really performant implementation by Martin Reinecke in ducc0 at https://gitlab.mpcdf.mpg.de/mtr/ducc/-/tree/ducc0 .

**Comment by hawkinsp** [comment]
I don't think we can use ducc0 ourselves in JAX, because it has a GPL license and JAX has an Apache license. However, nothing would stop someone from plugging it in as part of a separate library. We can walk someone through how to do that if interested. It would look very similar to the way PocketFFT is integrated into JAX.

A significant fraction of the time difference between JAX and SciPy can be remedied by `jit`-ing the fft function:

```
In [3]: %timeit sp_fft.fft(r)
603 ms ± 6.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [4]: %timeit jnp.fft.fft(r).block_until_ready()
768 ms ± 5.63 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [5]: import jax

In [6]: jfft = jax.jit(jnp.fft.fft)

In [7]: %timeit jfft(r).block_until_ready()
689 ms ± 7.64 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

This is actually something we can do for you in our standard library as a decorator on the `jnp.fft.fft` function; I'll send a PR doing that.

I'm not quite sure what accounts for the remaining time difference: it might be related to something like compiler options for example. I'll have to look further.

**Comment by Edenhofer** [comment]
That's a pity. Thanks, I would be very much interested in that!

I can't reproduce these numbers for `np.clomplex128`. The only reason I left out the jitted part is because it is as performant as the non-jitted part for me:
```text
POWER 22; DATA-TYPE <class 'numpy.complex128'>
NumPy: 0.1735294461427007s
SciPy: 0.13626850842826702s
JAX (unjitted): 0.2511234862857756s
JAX (jitted): 0.2408967502856935s
POWER 23; DATA-TYPE <class 'numpy.complex128'>
NumPy: 0.35449144728564924s
SciPy: 0.2941029569997876s
JAX (unjitted): 0.5531433257144401s
JAX (jitted): 0.5734982165714817s
POWER 24; DATA-TYPE <class 'numpy.complex128'>
NumPy: 0.7349333108572215s
SciPy: 0.5856138962855246s
JAX (unjitted): 1.260388970571512s
JAX (jitted): 1.2334323804285046s
```
These timings are of course excluding the time it took for JAX's jit to warm-up.

**Comment by mreineck** [comment]
> I don't think we can use ducc0 ourselves in JAX, because it has a GPL license and JAX has an Apache license. 

That's correct, but if we are talking about the FFT component only, I think I could make that available under BSD terms. Please let me know if you'd be interested in that.

(The directly FFT-related files in ducc0 still have the BSD licensing header, as far as I remember, but some of the support header files don't, so they'd need to be adjusted.)

**Comment by mreineck** [comment]
Just to clarify earlier parts of the discussion: the C version of pocketfft is the older one which should not be used anymore, unless the available environment is C only. The C++ version of pocketfft has better performance and many more capabilities and has the nice property of coming as a single header.
If you have access to C++17, I recommend the FFT component of ducc0, which is the evolution of C++ pocketfft with many minor improvements (but nothing really groundbreaking).

**Comment by Edenhofer** [comment]
Is there any interest in using the FFT component of ducc0 in JAX if it would have a BSD license?

If so, I would be happy to implement it.

**Comment by hawkinsp** [comment]
@Edenhofer @mreineck We would be very happy to use ducc0's FFT component, if it is available under an Apache/BSD license. We're really have two criteria here:
* fast
* under a liberal license (not more restrictive than JAX's own Apache license).

If Hartley transformations are useful (you tell me!) and there's a good implementation available that satisfies those criteria, they seem like something we could also plug into JAX.

I think we could mandate C++17 for JAX. We already require C++14.

**Comment by mreineck** [comment]
Concerning performance, here is a rough, qualitative  comparison to FFTW:
- for 1D transforms ducc0 is quite a bit slower (unless you include FFTW's planning overhead)
- for multi-D ducc0 is often faster, especially for single precision
- with increasing number of threads, ducc0 becomes slower compared to FFTW

In my eyes, the major selling point of ducc0 (apart from the licensing terms) would be the lack of explicit planning, which makes the interface really simple.

I'm not aware of another implementation which is licensed more liberally than GPL and has comparable functionality and performance - but if it exists, I'd love to hear about it!

Hartley transforms are really convenient in situations where you have real data and want your arrays to have exactly the same size in both nonharmonic and harmonic spaces. Whether this is attractive for many JAX users I can't really tell ;-)

**Comment by hawkinsp** [comment]
I gather @Edenhofer is quite interested in Hartley transforms :-)

**Comment by Edenhofer** [comment]
Since ducc0 is the successor to pocketfft, it should be a straightforward upgrade from pocketfft. I'll have a look at f58f1ee4563852de2853115d885c2eec7a69a77b and try to perform the upgrade.

Yes, I am definitely very interested in Hartley transformations. If me being interested in it is sufficient reason to include it in JAX, I will try to do that as well.

**Comment by mreineck** [comment]
I think the problem will be to provide Hartley transforms on GPU as well. It's probably not very much work, but someone (who knows about GPUs, i.e. not me...) has to care enough to do it.

**Comment by hawkinsp** [comment]
We could start by implementing a Hartley transform via Fourier transforms I think? (I'm not that familiar with the math here, perhaps there is some misunderstanding I have.) That's probably the right thing to do if you have an optimized Fourier transform GPU kernel but no Hartley transform.

**Comment by mreineck** [comment]
> We could start by implementing a Hartley transform via Fourier transforms I think?

Exactly, that's what ducc0 does as well. The post-processing is fairly minimal, but in multi-D transforms there are some tricky subtleties (see the last paragraph of http://www.fftw.org/fftw3_doc/Multi_002ddimensional-Transforms.html).

**Comment by Edenhofer** [comment]
Since Hartley transformations are linear just like Fourier transformations, their derivative should be simple enough to implement. I have no experience with writing code for the GPU though.

@mreineck As a first rough version, something like
```python
def hartley(a, axes=None):
    tmp = scipy.fft.fftn(a, axes=axes)
    return tmp.real + tmp.imag
```
should work for the GPU code, right?

**Comment by mreineck** [comment]
Yes, that should work fine. The drawback is that you generate a temporary complex array which is twice as large as your input.

Small remark:
>    `return tmp.real + tmp.imag`

As I recently noticed, this is not the generally used convention for the Hartley transform. It should actually read
`return tmp.real - tmp.imag`.
It doesn't change much in the grand scheme of things, but I'm currently updating ducc0 to support this convention as well, without breaking backwards compatibility.

**Comment by hawkinsp** [comment]
@mreineck we still require C++14 because there are some C++17 problems for Windows/MSVC that TensorFlow hasn't been able to resolve yet and we share TF's compiler version policy. That said, I'm not sure they are actually problems for JAX, so if it's a big problem we might be able to bump our compiler requirements sooner than TF does.

The other pieces sound fine!

**Comment by hawkinsp** [comment]
@mreineck I asked about C++17. It's currently blocked on a compiler upgrade, but we're hoping "next week" maybe.

(It's harder to do this than you might think, because we have to ship `manylinux2010` or `manylinux2014` wheels, and the problem of having a C++17 toolchain that targets those specifications is probably apparent from the names... so building our wheels involves a custom toolchain.)

**Comment by mreineck** [comment]
Great news, thanks!
(I know that this kind of update is hard ... `scipy` can't do it yet either due to manylinux constraints.)

**Comment by mreineck** [comment]
I noticed that JAX is not currently making use of multi-threading in pocketfft - this could be achieved by passing an additional `nthreads` argument (defaulted to 1).
Is this deliberate? If yes, we can probably simplify the new code by disabling multi-threaduing support via a preprocessor macro.

**Comment by hawkinsp** [comment]
@Edenhofer Is there still more for us to do here?

**Comment by Edenhofer** [comment]
My hope is that the `-ffast-math` compiler flag for ducc0 resolves this problem but I was planning to rerun the experiment just to be sure. I'll report back next week :)

**Comment by Edenhofer** [comment]
Fixed with #12122 .
